nengo/builder/tests/test_transforms.py

```diff
@@ -109,9 +109,169 @@ def test_convinc_2d(
+def test_convtransposeinc_2d(
+    channels_last, stride0, stride1, kernel0, kernel1, padding, rng, allclose, plt
+):
+    """Test ConvTransposeInc by ensuring it is the transpose of ConvInc.
+
+    Since convolution is a linear operator, it can be expressed as a matrix ``A``.
+    We can therefore state that ``C.dot(A.dot(x)) == (A.T.dot(C.T)).T.dot(x)``,
+    for an arbitrary vector ``x`` and arbitrary matrix ``C``.
+
+    This test asserts this identity, and thereby tests the ``ConvTransposeInc`` operator
+    against the ``ConvInc`` operator.
+    """
+    shape0 = 16
+    shape1 = 17
+    in_channels = 32
+    out_channels = 64
+
+    x_shape = (
+        (shape0, shape1, in_channels)
+        if channels_last
+        else (in_channels, shape0, shape1)
+    )
+
+    conv = Convolution(
+        out_channels,
+        x_shape,
+        kernel_size=(kernel0, kernel1),
+        strides=(stride0, stride1),
+        padding=padding,
+        channels_last=channels_last,
+    )
+
+    nk = 10  # number of vectors we test ConvTransposeInc with
+    C = rng.randn(nk, conv.output_shape.size)
+
+    # compute ``conv_output = C.dot(A.dot(x))``, where ``A`` is the convolution operator
+    x = Signal(rng.randn(*x_shape))
+    w = Signal(rng.randn(kernel0, kernel1, in_channels, out_channels))
+    y = Signal(np.zeros(conv.output_shape.shape))
+
+    signals = {sig: np.array(sig.initial_value) for sig in (x, w, y)}
+    step_conv = ConvInc(w, x, y, conv).make_step(signals, None, None)
+    step_conv()
+
+    x_flat = signals[x].ravel()
+    y_flat = signals[y].ravel()
+    conv_output = C.dot(y_flat)
+
+    # compute ``conv_transpose_output = (A.T.dot(C.T)).T.dot(x)``, where ``A.T`` is the
+    # transpose convolution operator (applied to one column of ``C.T`` at a time)
+    conv_transpose = ConvolutionTranspose(
+        in_channels,
+        conv.output_shape,
+        output_shape=x_shape,
+        kernel_size=(kernel0, kernel1),
+        strides=(stride0, stride1),
+        padding=padding,
+        channels_last=channels_last,
+    )
+    xt = Signal(np.zeros(conv_transpose.input_shape.shape))
+    wt = Signal(np.transpose(w.initial_value, (0, 1, 3, 2)))
+    yt = Signal(np.zeros(conv_transpose.output_shape.shape))
+
+    signals_tr = {sig: np.array(sig.initial_value) for sig in (xt, wt, yt)}
+    step_trans = ConvTransposeInc(wt, xt, yt, conv_transpose).make_step(
+        signals_tr, None, None
+    )
+
+    AtCt = []
+    for k in range(nk):
+        signals_tr[xt][:] = C[k].reshape(conv_transpose.input_shape.shape)
+        signals_tr[yt][:] = 0
+        step_trans()
+        AtCt.append(signals_tr[yt].copy().ravel())
+
+    AtCt = np.array(AtCt)
+    conv_transpose_output = AtCt.dot(x_flat)
+    assert conv_transpose_output.shape == conv_output.shape
+
+    # Check whether the outputs match. If they don't, print/plot some diagnostic output.
+    success = allclose(conv_transpose_output, conv_output)
+    if not success:
```

Comment:
This seems like a lot of code just to diagnose test failures (which seems like it'd be of pretty limited utility after this PR is merged). If we don't want to lose this code snippet I'd be inclined to just add it as a comment on the PR (and then just `assert success` here).

Reply:
If we're having it as a comment here, I don't see the benefit of that versus having it be active code. It's something that I think I tried to get rid of more than once, and then found another bug and realized I needed it again. It can be a real pain to diagnose these bugs, and there's no guarantees more won't come up in the future (there are so many parameters, and thus so many edge cases).

Reply:
It just seems odd to me that almost half the code in this test is not actually part of the test, it's just there to help with debugging in case the test fails. That's certainly not something we've typically done throughout the code base.